### ISSUE TITLE
HTML tags misinterpreted as beamer optional arguments

### ISSUE THREAD

**Issue by yihui** [jgm/pandoc#3368]
I just noticed a change in my LaTeX output after I updated Pandoc from 1.18 to 1.19. For Pandoc <= 1.18, HTML tags will be ignored in the LaTeX output, e.g. the Markdown input

    \textsf<span>foo</span>

will be converted to

    \textsf{foo}

in .tex output, i.e. the HTML tags will be stripped off. I liked this (probably undocumented) feature because the same Markdown source file can be converted to either HTML or LaTeX. When we convert to one format, the commands/tags specific to the other format will be discarded.

With Pandoc >= 1.19, we get

    \textsf<span>foo

in .tex output, which does not make sense.

I took a look at the commits and I guess 8d7ecc27a19facf6c4b50a65ee1029c105aacdb2 introduced the change. I wonder if it is possible to preserve the original behavior, although it may be an unintended feature.

**Comment by ickc** [comment]
It's because of this:

```bash
# pandoc 1.19.1
printf "%s\n\n" '\textsf<span>foo</span>' | pandoc -f markdown -t native
[Para [RawInline (Format "tex") "\\textsf<span>",Str "foo",RawInline (Format "html") "</span>"]]
# pandoc 1.18
printf "%s\n\n" '\textsf<span>foo</span>' | ~/Downloads/pandoc-1.18-osx/local/bin/pandoc  -f markdown -t native
[Para [RawInline (Format "tex") "\\textsf",Span ("",[],[]) [Str "foo"]]]
```

It's hard to say what should be expected, because of the mingling of raw TeX and raw HTML. And in fact this use looks like a hack to me (which fortunately works in the past). A better approach would be to always do it in a span with a certain attribute, and then use a filter to convert that native span with that attribute into raw TeX when the output is LaTeX. The bonus here is that within the span there can be markdown (no markdown allowed in raw TeX).

But on the other hand, this still seems like a bug for a cornering case (because it breaks the html apart).

**Comment by yihui** [comment]
It is definitely a hack, which I borrowed from #2453.

**Comment by ickc** [comment]
Parsing raw LaTeX in markdown source is actually quite complicated, and currently has quite a few bugs around it (one I can search right now is #1390).

A hack is a quick fix, but not long term solution. A hack relies on an undocumented, untested behavior (meaning such behavior is not in the tests of the repository), so in the future it can disappeared silently without being noticed (by Travis and auto-tests).

I think the much better and stable way to do is via filter. There's a few mentioned in [Pandoc Filters · jgm/pandoc Wiki](https://github.com/jgm/pandoc/wiki/Pandoc-Filters), e.g.

- [pandocfilters/latexdivs.py at master · jgm/pandocfilters](https://github.com/jgm/pandocfilters/blob/master/examples/latexdivs.py)
- [Pandoc filter to insert arbitrary raw output markup as Code/CodeBlocks with an attribute raw=<outputformat>.](https://gist.github.com/bpj/e6e53cbe679d3ec77e25)

In principle, some sort of `latexspans.py` can be created which uses a native span and convert to inline LaTeX macros. But there seems complication around `{...}` or not. (see #1390)

Again, you get markdown in LaTeX environments for free as a side-benefit.

**Comment by yihui** [comment]
Yes, I definitely understand the consequences of using undocumented "features", and tend to agree that filters will be a much better way to go. I'll learn more about them when I have time. For now, this issue is not very urgent, since our production environment will use Pandoc 1.17.2 for quite a while. Thanks!

**Comment by jgm** [comment]
This was really an unintended consequence.  I knew there was *some* reason why I'd been initially reluctant to support the angle-bracketed beamer options, and this was it, I guess.

Is there any reasonable way to limit what we allow in the angle brackets?  Are options of that kind used anywhere else besides beamer?  Do they always have numerical contents?

**Comment by yihui** [comment]
From my experience, LaTeX commands are typically followed by square brackets (options) or curly braces. I have only seen the use of angle brackets in beamer. I think the possible content is numeric plus commas `,` and dashes `-`, which won't be valid HTML tags. Perhaps you can be more conservative to take a look at the content inside `<>` before deciding what to do (e.g. keep it only if it is `[-,0-9]`).

**Comment by jgm** [comment]
Or we could just check to see if it's a valid HTML tag before treating it as a special LaTeX option.

**Comment by yihui** [comment]
Yep, that will also work.

**Comment by ickc** [comment]
More about filters:

Hi, @yihui, didn't recognize you as the Knitr guy before. Currently pandoc filters can be written in [a few different languages](https://github.com/jgm/pandoc/wiki/Pandoc-Filters). But that doesn't seem to fit the Knitr picture very well. You may as well develop another interface for people to write filters in R. I guess that would be very interesting for your audiences.

If that's not your intension and the existing languages for writing filters seems too complicated for the workflow, `@tarleb` is developing [tarleb/panlunatic: Custom json writer for portable pandoc filtering without external programs](https://github.com/tarleb/panlunatic), which basically uses the embedded lua interpreter in pandoc to interpret filters written in lua.

**Comment by yihui** [comment]
@ickc Thanks for all the helpful tips! I will definitely learn the filters when I have time.

**Comment by yihui** [comment]
@jgm Is there a chance to get this issue fixed in the next couple of months? Thanks!

**Comment by jgm** [comment]
@yihui  I'll add it to the pandoc-2.0 milestone.

**Comment by jgm** [comment]
@yihui in b98a05d604ab4353c2e684beecf669b182d88906 I tighten up parsing rule for beamer-style `<>` arguments.

But this doesn't make your original case work.
Instead we now get:
```
 % pandoc -t native
\textsf<span>foo</span>
[Para [RawInline (Format "tex") "\\textsf<",Str "span>foo",RawInline (Format "html") "</span>"]]
```
Note that the initial `<` is interpreted as an argument of `\textsf` -- as it should be, that's exactly how it is interpreted by LaTeX!

Then I reverted the commit, because I realized that beamer angle arguments can contain lots of things, even letters.

I think you're going to have to switch over to using filters, rather than relying on this hack.
